File: packages/contracts/rbac/rbac.go
package rbac

// Platform roles from TZ §24 (items 1–6 implemented in this slice).
const (
	RoleGuest         = "guest"
	RoleStudent       = "student"
	RoleGuardian      = "guardian"
	RoleCoach         = "coach"
	RoleRenter        = "renter"
	RoleAdministrator = "administrator"
	// Legacy local role — same privileges as administrator.
	RolePlatformAdmin = "platform_admin"
)

type Permission string

const (
	PermUsersRead         Permission = "users.read"
	PermUsersCreate       Permission = "users.create"
	PermUsersUpdate       Permission = "users.update"
	PermUsersDelete       Permission = "users.delete"
	PermContentRead       Permission = "content.read"
	PermContentWrite      Permission = "content.write"
	PermContentDelete     Permission = "content.delete"
	PermSchoolRead        Permission = "school.read"
	PermAttendanceConfirm Permission = "attendance.confirm"
	PermDependantsRead    Permission = "dependants.read"
	PermHallsRead         Permission = "halls.read"
)

var allPermissions = []Permission{
	PermUsersRead,
	PermUsersCreate,
	PermUsersUpdate,
	PermUsersDelete,
	PermContentRead,
	PermContentWrite,
	PermContentDelete,
	PermSchoolRead,
	PermAttendanceConfirm,
	PermDependantsRead,
	PermHallsRead,
}

var rolePermissions = map[string][]Permission{
	RoleStudent: {
		PermSchoolRead,
	},
	RoleGuardian: {
		PermDependantsRead,
		PermSchoolRead,
	},
	RoleCoach: {
		PermSchoolRead,
		PermAttendanceConfirm,
	},
	RoleRenter: {
		PermHallsRead,
	},
	RoleAdministrator: allPermissions,
	RolePlatformAdmin:   allPermissions,
}

// NormalizeRole maps empty/unknown values to student for authenticated principals.
func NormalizeRole(role string) string {
	if role == "" {
		return RoleStudent
	}
	if IsKnownRole(role) {
		return role
	}
	return RoleStudent
}

func IsKnownRole(role string) bool {
	switch role {
	case RoleGuest, RoleStudent, RoleGuardian, RoleCoach, RoleRenter, RoleAdministrator, RolePlatformAdmin:
		return true
	default:
		return false
	}
}

func IsAdministrator(role string) bool {
	role = NormalizeRole(role)
	return role == RoleAdministrator || role == RolePlatformAdmin
}

func PermissionsFor(role string) map[Permission]bool {
	role = NormalizeRole(role)
	grants := rolePermissions[role]
	if grants == nil {
		grants = rolePermissions[RoleStudent]
	}
	out := make(map[Permission]bool, len(grants))
	for _, p := range grants {
		out[p] = true
	}
	return out
}

func HasPermission(role string, perm Permission) bool {
	return PermissionsFor(role)[perm]
}

func CanManageUsers(role string) bool {
	role = NormalizeRole(role)
	return HasPermission(role, PermUsersCreate) &&
		HasPermission(role, PermUsersUpdate) &&
		HasPermission(role, PermUsersDelete)
}

func AssignableRoles() []string {
	return []string{
		RoleStudent,
		RoleGuardian,
		RoleCoach,
		RoleRenter,
		RoleAdministrator,
	}
}

var rolePriority = []string{
	RolePlatformAdmin,
	RoleAdministrator,
	RoleCoach,
	RoleGuardian,
	RoleRenter,
	RoleStudent,
}

// NormalizeRolesList deduplicates and drops unknown roles.
func NormalizeRolesList(roles []string) []string {
	if len(roles) == 0 {
		return []string{RoleStudent}
	}
	seen := map[string]struct{}{}
	out := make([]string, 0, len(roles))
	for _, role := range roles {
		n := NormalizeRole(role)
		if n == RoleGuest {
			continue
		}
		if _, ok := seen[n]; ok {
			continue
		}
		seen[n] = struct{}{}
		out = append(out, n)
	}
	if len(out) == 0 {
		return []string{RoleStudent}
	}
	return out
}

// PrimaryRole picks the highest-priority role for routing/display.
func PrimaryRole(roles []string) string {
	roles = NormalizeRolesList(roles)
	set := map[string]struct{}{}
	for _, role := range roles {
		set[role] = struct{}{}
	}
	for _, role := range rolePriority {
		if _, ok := set[role]; ok {
			return role
		}
	}
	return RoleStudent
}

func MergePermissions(roles []string) map[Permission]bool {
	out := map[Permission]bool{}
	for _, role := range NormalizeRolesList(roles) {
		for perm, ok := range PermissionsFor(role) {
			if ok {
				out[perm] = true
			}
		}
	}
	return out
}

func HasPermissionForRoles(roles []string, perm Permission) bool {
	return MergePermissions(roles)[perm]
}

func HasRole(roles []string, role string) bool {
	role = NormalizeRole(role)
	for _, item := range NormalizeRolesList(roles) {
		if item == role {
			return true
		}
	}
	return false
}

func IsAdministratorInRoles(roles []string) bool {
	return HasRole(roles, RoleAdministrator) || HasRole(roles, RolePlatformAdmin)
}

File: apps/web/src/lib/rbac.ts
export const ROLES = {
  guest: "guest",
  student: "student",
  guardian: "guardian",
  coach: "coach",
  renter: "renter",
  administrator: "administrator",
  platform_admin: "platform_admin",
} as const;

export type UserRole =
  | typeof ROLES.student
  | typeof ROLES.guardian
  | typeof ROLES.coach
  | typeof ROLES.renter
  | typeof ROLES.administrator
  | typeof ROLES.platform_admin;

export type Permission =
  | "users.read"
  | "users.create"
  | "users.update"
  | "users.delete"
  | "content.read"
  | "content.write"
  | "content.delete"
  | "school.read"
  | "attendance.confirm"
  | "dependants.read"
  | "halls.read";

const ALL_PERMISSIONS: Permission[] = [
  "users.read",
  "users.create",
  "users.update",
  "users.delete",
  "content.read",
  "content.write",
  "content.delete",
  "school.read",
  "attendance.confirm",
  "dependants.read",
  "halls.read",
];

const ROLE_PERMISSIONS: Record<UserRole, Permission[]> = {
  student: ["school.read"],
  guardian: ["dependants.read", "school.read"],
  coach: ["school.read", "attendance.confirm"],
  renter: ["halls.read"],
  administrator: ALL_PERMISSIONS,
  platform_admin: ALL_PERMISSIONS,
};

export const ROLE_LABELS: Record<UserRole, string> = {
  student: "Ученик",
  guardian: "Опекун",
  coach: "Тренер",
  renter: "Арендатор",
  administrator: "Администратор",
  platform_admin: "Администратор платформы",
};

export const ASSIGNABLE_ROLES: UserRole[] = [
  "student",
  "guardian",
  "coach",
  "renter",
  "administrator",
];

export function normalizeRole(raw: unknown): UserRole {
  const value = typeof raw === "string" ? raw : "";
  if (value === "administrator" || value === "platform_admin") return value;
  if (value === "guardian" || value === "coach" || value === "renter") return value;
  return "student";
}

export function normalizeRoles(raw: unknown): UserRole[] {
  if (Array.isArray(raw)) {
    const roles = raw.map((item) => normalizeRole(item)).filter((role, index, list) => list.indexOf(role) === index);
    return roles.length > 0 ? roles : ["student"];
  }
  return [normalizeRole(raw)];
}

export function permissionsForRoles(roles: UserRole[]): Permission[] {
  const merged = new Set<Permission>();
  for (const role of normalizeRoles(roles)) {
    for (const perm of permissionsFor(role)) {
      merged.add(perm);
    }
  }
  return [...merged];
}

export function hasPermissionForRoles(roles: UserRole[], permission: Permission): boolean {
  return permissionsForRoles(roles).includes(permission);
}

export function hasRole(roles: UserRole[], role: UserRole): boolean {
  return normalizeRoles(roles).includes(role);
}

export function isAdminPrincipal(roles: UserRole[] | UserRole): boolean {
  const list = Array.isArray(roles) ? normalizeRoles(roles) : [normalizeRole(roles)];
  return list.includes("administrator") || list.includes("platform_admin");
}

export function canManageUsers(roles: UserRole[] | UserRole): boolean {
  const list = Array.isArray(roles) ? roles : [roles];
  return hasPermissionForRoles(list, "users.create") && hasPermissionForRoles(list, "users.update") && hasPermissionForRoles(list, "users.delete");
}

export function homePathForRoles(roles: UserRole[], profileReady: boolean): string {
  const list = normalizeRoles(roles);
  if (isAdminPrincipal(list)) return "/admin";
  if (hasRole(list, "coach")) return "/coach";
  if (hasRole(list, "guardian")) return "/guardian";
  if (hasRole(list, "renter")) return "/renter";
  return profileReady ? "/profile" : "/onboarding";
}

export function primaryRole(roles: UserRole[]): UserRole {
  const priority: UserRole[] = ["platform_admin", "administrator", "coach", "guardian", "renter", "student"];
  const list = normalizeRoles(roles);
  for (const role of priority) {
    if (list.includes(role)) return role;
  }
  return "student";
}

export function permissionsFor(role: UserRole): Permission[] {
  return ROLE_PERMISSIONS[role] ?? ROLE_PERMISSIONS.student;
}

export function hasPermission(role: UserRole, permission: Permission): boolean {
  return permissionsFor(role).includes(permission);
}

/** @deprecated use isAdminPrincipal(roles) */
export function isAdminPrincipalRole(role: UserRole): boolean {
  return role === "administrator" || role === "platform_admin";
}

/** @deprecated use canManageUsers(roles) */
export function canManageUsersRole(role: UserRole): boolean {
  return hasPermission(role, "users.create") && hasPermission(role, "users.update") && hasPermission(role, "users.delete");
}

export function homePathForRole(role: UserRole, profileReady: boolean): string {
  return homePathForRoles([role], profileReady);
}

export function formatRoles(roles: UserRole[]): string {
  return normalizeRoles(roles)
    .map((role) => ROLE_LABELS[role] ?? role)
    .join(", ");
}
